src/main/java/pl/oskarkowalski/recruitment/ticketbookingapp/services/ReservationService.java

```diff
@@ -0,0 +1,171 @@
+package pl.oskarkowalski.recruitment.ticketbookingapp.services;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import pl.oskarkowalski.recruitment.ticketbookingapp.dtos.ReservationDto;
+import pl.oskarkowalski.recruitment.ticketbookingapp.dtos.TicketDto;
+import pl.oskarkowalski.recruitment.ticketbookingapp.entities.Reservation;
+import pl.oskarkowalski.recruitment.ticketbookingapp.entities.Screening;
+import pl.oskarkowalski.recruitment.ticketbookingapp.entities.Seat;
+import pl.oskarkowalski.recruitment.ticketbookingapp.entities.Ticket;
+import pl.oskarkowalski.recruitment.ticketbookingapp.enums.SeatRow;
+import pl.oskarkowalski.recruitment.ticketbookingapp.enums.TicketPrice;
+import pl.oskarkowalski.recruitment.ticketbookingapp.exceptions.ResourceNotFoundException;
+import pl.oskarkowalski.recruitment.ticketbookingapp.repositories.ScreeningRepository;
+import pl.oskarkowalski.recruitment.ticketbookingapp.repositories.TicketRepository;
+
+import java.time.LocalDateTime;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Service
+public class ReservationService {
+
+  private final ScreeningRepository screeningRepository;
+  private final TicketRepository ticketRepository;
+
+  @Value("${discount.code}")
+  private char[] discountCode;
+
+  @Autowired
+  public ReservationService(ScreeningRepository screeningRepository, TicketRepository ticketRepository) {
+    this.screeningRepository = screeningRepository;
+    this.ticketRepository = ticketRepository;
+  }
+
+  public boolean isOkayToBuy(ReservationDto reservationDto) {
+    TicketDto ticketDto = reservationDto.getTickets().iterator().next();
+    Screening screening = screeningRepository.findById(ticketDto.getScreeningId())
+        .orElseThrow(() -> new ResourceNotFoundException("Screening", ticketDto.getScreeningId()));
+    return isThereNoSingleFreeSeatOnScreening(reservationDto.getTickets(), screening) && isThisAGoodTime(reservationDto.getOrderTime(), screening);
+  }
+
+  public boolean isThisAGoodTime(LocalDateTime orderTime, Screening screening) {
+    LocalDateTime startTime = screening.getTime();
+    return orderTime.compareTo(startTime.minusMinutes(15)) <= 0;
+  }
+
+  public Boolean isDiscountCodeValid(String providedCode) {
+    char[] providedCodeLetters = providedCode.toCharArray();
+    return Arrays.equals(providedCodeLetters, discountCode);
+  }
+
+  public Double calculatePrice(Reservation reservation) {
+    Double totalPrice = (double) 0;
+    List<TicketPrice> ticketPrices = reservation.getTickets().stream().map(Ticket::getPrice).collect(Collectors.toList());
+
+    for (TicketPrice price : ticketPrices) {
```

Comment:
nie da się tego napisać bardziej w stylu javy>=8 ?

Reply:
Jest kilka opcji do zrobienia w tym przypadku - najłatwiejsza to zmapowanie do doubli i zsumowanie. Jest to do poprawy przy okazji konwersji owych Doubli na BigDecimale.